### PR TITLE
Add repeat method on slice

### LINKED ISSUE #48784: `repeat` method for (byte) slices?
There exists a [`str::repeat`](https://doc.rust-lang.org/std/primitive.str.html#method.repeat) method, but no equivalent for (byte) slices. This would seem like a useful feature. Should we consider adding it to the stdlib?

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #48784.

### CODE REVIEW

File: src/liballoc/slice.rs [hunk 1]
```diff
@@ -1722,6 +1722,54 @@ impl<T> [T] {
...
+        // `n = 2^expn + rem (2^expn > rem, expn >= 0, rem >= 0)`.
+        // `2^expn` is the number represented by the leftmost '1' bit of `n`,
+        // and `rem` is the remaining part of `n`.
+
+        // Using `Vec` to access `set_len()`.
+        let mut buf = Vec::with_capacity(self.len() * n);
+
+        // `2^expn` repetition is done by doubling `buf` `expn`-times.
+        buf.extend(self);
+
```
Comment (sinkuu) [inline src/liballoc/slice.rs]: You have missed this part: https://github.com/sinkuu/rust/blob/3d58543d49266a7ec3eb5f5f2ffaf902fce17c53/src/liballoc/str.rs#L2084-L2102

**Comment by sinkuu** [comment]
Can you replace the implementation of `str::repeat` with  `String::from_utf8_unchecked(self.as_bytes().repeat(n))` to avoid code duplication?

**Comment by GuillaumeGomez** [comment]
@sinkuu: That's what I thought about doing but they seem to make specific `str` operations (checking if they're at the end of a char, etc...). Did I misunderstood?

**Comment by sinkuu** [comment]
@GuillaumeGomez Valid `String`s must have char boundary at their ends, so repeating them does not require special considerations.
Just like `String::push_str` simply extends their internal `Vec<u8>` with the other.
https://doc.rust-lang.org/1.22.0/src/alloc/string.rs.html#804-808

File: src/liballoc/slice.rs [hunk 2]
```diff
@@ -1722,6 +1722,73 @@ impl<T> [T] {
         // NB see hack module in this file
         hack::into_vec(self)
     }
+
+    /// Creates a [`Vec`] by repeating a slice `n` times.
+    ///
+    /// [`Vec`]: ../../std/vec/struct.Vec.html
```
Comment (scottmcm) [inline src/liballoc/slice.rs]:
```
[01:26:06] std/primitive.slice.html:1018: broken link - /checkout/obj/build/x86_64-unknown-linux-gnu/std/vec/struct.Vec.html
[01:26:20] thread 'main' panicked at 'found some broken links', tools/linkchecker/main.rs:49:9
```

File: src/liballoc/slice.rs [hunk 3]
```diff
@@ -1722,6 +1722,73 @@ impl<T> [T] {
...
+
+                m >>= 1;
+            }
+        }
+
+        // `rem` (`= n - 2^expn`) repetition is done by copying
+        // first `rem` repetitions from `buf` itself.
+        let rem_len = self.len() * n - buf.len(); // `self.len() * rem`
+        if rem_len > 0 {
+            // `buf.extend(buf[0 .. rem_len])`:
```
Comment (Kimundi) [inline src/liballoc/slice.rs]: Hm, why is the code here not just literally `buf.extend(&buf[..rem_len])`? Worse codegen?

File: src/liballoc/slice.rs [hunk 4]
```diff
@@ -1722,6 +1722,73 @@ impl<T> [T] {
...
+    pub fn repeat(&self, n: usize) -> Vec<T> where T: Copy {
+        if n == 0 {
+            return Vec::new();
+        }
+
+        // If `n` is larger than zero, it can be split as
+        // `n = 2^expn + rem (2^expn > rem, expn >= 0, rem >= 0)`.
+        // `2^expn` is the number represented by the leftmost '1' bit of `n`,
+        // and `rem` is the remaining part of `n`.
+
```
Comment (Kimundi) [inline src/liballoc/slice.rs]: It took me a few minutes to figure out why this function is so complex - maybe add a comment that explicitly explains that its trying to minimize the `copy_nonoverlapping` calls by always copying as large chunks as possible?

File: src/liballoc/slice.rs (hunk 3 above)
Comment (sinkuu) [reply src/liballoc/slice.rs]: Because you cannot have `&mut buf` and `&buf` at the same time. The comment was meant to be like a pseudocode.

Comment (Kimundi) [reply src/liballoc/slice.rs]: Ah, of course. I somehow forgot about the borrow checker for a minute there, too much C++ recently...

File: src/liballoc/slice.rs (hunk 4 above)
Comment (GuillaumeGomez) [reply src/liballoc/slice.rs]: I just moved the code around, didn't modify it. It's exactly like it was before in `str`.

**Comment by GuillaumeGomez** [comment]
It passed tests. Is there anything else I should do?

File: src/liballoc/slice.rs [hunk 5]
```diff
@@ -1722,6 +1722,71 @@ impl<T> [T] {
...
+    /// Creates a vector by repeating a slice `n` times.
+    ///
+    /// # Examples
+    ///
+    /// Basic usage:
+    ///
+    /// ```
+    /// assert_eq!([1, 2].repeat(3), vec![1, 2, 1, 2, 1, 2]);
+    /// ```
+    #[stable(feature = "repeat_slice", since = "1.27.0")]
```
Comment (Kimundi) [inline src/liballoc/slice.rs]: All new APIs should start off as unstable. Just invent a new feature name for it, like "repeat_generic_slice" or so.

**Review by Kimundi — CHANGES_REQUESTED** [review]

File: src/liballoc/slice.rs (hunk 5 above)
Comment (GuillaumeGomez) [reply src/liballoc/slice.rs]: Ok!

**Comment by GuillaumeGomez** [comment]
Updated.

**Comment by Kimundi** [comment]
Ah, sorry I didn't notice it before, but if you could add some basic tests that check that `repeat` works as expected on a few different slice types that would be good. (Even if its the same code, just moved around, there might always be something unexpected going on)

**Comment by GuillaumeGomez** [comment]
Sure. I'll add them later in the week.

**Comment by GuillaumeGomez** [comment]
Added a test.

File: src/liballoc/slice.rs [hunk 6]
```diff
@@ -1722,6 +1722,73 @@ impl<T> [T] {
...
         hack::into_vec(self)
     }
+
+    /// Creates a vector by repeating a slice `n` times.
+    ///
+    /// # Examples
+    ///
+    /// Basic usage:
+    ///
+    /// ```
```
Comment (scottmcm) [inline src/liballoc/slice.rs]:
```rust
[01:10:05] ---- slice.rs - slice::[T]::repeat (line 1732) stdout ----
[01:10:05] 	error[E0658]: use of unstable library feature 'repeat_generic_slice': it's on str, why not on slice? (see issue #48784)
[01:10:05]  --> slice.rs:1733:19
[01:10:05]   |
[01:10:05] 4 | assert_eq!([1, 2].repeat(3), vec![1, 2, 1, 2, 1, 2]);
[01:10:05]   |                   ^^^^^^
[01:10:05]   |
[01:10:05]   = help: add #![feature(repeat_generic_slice)] to the crate attributes to enable
```

**Review by scottmcm — COMMENTED** [review]
BTW, I mentioned this on IRC yesterday, and someone tried it out and got really excited that it was like 90x faster than what they were doing before 😄

**Comment by GuillaumeGomez** [comment]
Fixed doc example.

**Comment by GuillaumeGomez** [comment]
cc @Kimundi

**Comment by pietroalbini** [comment]
Ping from triage @Kimundi! The author pushed new commits, can you review this PR again?

**Comment by GuillaumeGomez** [comment]
Anyone?

**Comment by pietroalbini** [comment]
Can someone from @rust-lang/libs review this PR? Thanks!

**Comment by pietroalbini** [comment]
Ping from triage! Can @Kimundi (or someone else from @rust-lang/libs) review this?

**Comment by GuillaumeGomez** [comment]
I can't see which test failed. Just that we got a signal 11...

**Comment by Mark-Simulacrum** [comment]
Possibly a resurgence of https://github.com/rust-lang/rust/issues/49775? cc @alexcrichton 

But probably a spurious failure, seems somewhat unlikely to be caused by these changes.

@bors retry

**Pull request merged**